x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java

```diff
@@ -96,9 +97,13 @@ private void preFlight(final Operation operation) {
     }
 
     @Override
-    public NoOpResult noOp(NoOp noOp) {
-        // TODO: Make sure we process NoOp once.
-        return super.noOp(noOp);
+    protected Optional<Exception> preFlightCheckForNoOp(NoOp noOp) {
+        if (noOp.origin() == Operation.Origin.PRIMARY && hasBeenProcessedBefore(noOp)) {
+            // See the comment in #indexingStrategyForOperation for the explanation why we can safely skip this operation.
+            return Optional.of(new AlreadyProcessedFollowingEngineException(shardId, noOp.seqNo()));
```

Comment:
Note that this will require changing the other PR to accept no opes with seq > gcp having no term and replicating them with current term.

Reply:
Yes, will do.

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/34408/commits/8f193929a35e52b5a80f98fe38df75548abb78ee